cassiopeia/simulator/SequentialLineageTracingDataSimulator.py

```diff
@@ -0,0 +1,280 @@
+        if initiation_rate <= 0:
+            raise DataSimulatorError("Initiation rate must be positive.")
+        if not isinstance(continuation_rate, float):
+            raise DataSimulatorError("Continuation rate must be a float.")
+        if continuation_rate <= 0:
+            raise DataSimulatorError("Continuation rate must be positive.")
+        if not isinstance(state_priors, dict):
+            raise DataSimulatorError(
+                "State priors dictionary is required" )
+        if np.any(np.array(list(state_priors.values())) < 0):
+            raise DataSimulatorError(
+                "State prior to be non-negative."
+            )
+        Z = np.sum([v for v in state_priors.values()])
+        if not math.isclose(Z, 1.0):
+            raise DataSimulatorError("State priors do not sum to 1.")
+
+
+        self.size_of_cassette = size_of_cassette
+        self.number_of_cassettes = number_of_cassettes
+
+        self.initiation_rate = initiation_rate
+        self.continuation_rate = continuation_rate
+        self.state_priors = state_priors
+
+        self.heritable_silencing_rate = heritable_silencing_rate
+        self.stochastic_silencing_rate = stochastic_silencing_rate
+
+        self.heritable_missing_data_state = heritable_missing_data_state
+        self.stochastic_missing_data_state = stochastic_missing_data_state
+
+        self.random_seed = random_seed
+
+    def overlay_data(self, tree: CassiopeiaTree):
+        """Overlays Cas9-based lineage tracing data onto the CassiopeiaTree.
+
+        Args:
+            tree: Input CassiopeiaTree
+        """
+
+        if self.random_seed is not None:
+            np.random.seed(self.random_seed)
+
+        # initialize character matrix 
+        number_of_characters = self.number_of_cassettes * self.size_of_cassette
+        initiation_sites = range(0, number_of_characters, self.size_of_cassette)
+        character_matrix = {}
+        for node in tree.nodes:
+            character_matrix[node] = [-1] * number_of_characters
+
+        for node in tree.depth_first_traverse_nodes(tree.root, postorder=False):
+
+            if tree.is_root(node):
+                character_matrix[node] = [0] * number_of_characters
+                continue
+
+            parent = tree.parent(node)
+            life_time = tree.get_time(node) - tree.get_time(parent)
+
+            character_array = character_matrix[parent].copy()
+            new_edits = []
+
+            # sequentially edit each cassette
+            for c in initiation_sites:
+                # missing
+                if character_array[c] == -1:
+                    continue
+                # initiated
+                elif character_array[c] != 0:
+                    time = 0
+                    i = c + 1
+                    while time < life_time and i < c + self.size_of_cassette:
+                        if character_array[i] == 0:
+                            time = time + (-np.log(np.random.uniform())
```

Comment:
sorry again if I misunderstand the math, but couldn't this also just be drawn from an exponential as below, `time += np.random.exponential(1/self.continuation_rate)`?  Or is this a different waiting time distribution?

Reply:
Thanks! Missed that one